lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,419 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* global document, getComputedStyle */
+
+const Gatherer = require('../gatherer');
+const pageFunctions = require('../../../lib/page-functions.js');
+const TARGET_SELECTORS = [
+  'button',
+  'a',
+  'input',
+  'textarea',
+  'select',
+  'option',
+  '[role=button]',
+  '[role=checkbox]',
+  '[role=link]',
+  '[role=menuitem]',
+  '[role=menuitemcheckbox]',
+  '[role=menuitemradio]',
+  '[role=option]',
+  '[role=scrollbar]',
+  '[role=slider]',
+  '[role=spinbutton]',
+];
+
+// /**
+//  * @param {Element} node
+//  * @param {CSSStyleDeclaration} style
+//  * @returns {{left: number, top: number, bottom: number, right: number, width: number, height: number}}
+//  */
+// function getInteractiveBCR(node, style = getComputedStyle(node)) {
+//   const {display, overflowX, overflowY} = style;
+
+//   if (display === 'inline' && overflowX === 'visible' && overflowY === 'visible') {
+//     const range = document.createRange();
+//     range.selectNodeContents(node);
+
+//     return range.getBoundingClientRect();
+//   }
+
+//   return node.getBoundingClientRect();
+// }
+
+function allBCRsEmpty(bcrs) {
+  return (
+    bcrs.length === 0 || bcrs.every(bcr => bcr.width === 0 && bcr.height === 0)
+  );
+}
+
+/**
+ * @param {Element} node
+ */
+function isVisible({node, bcrs = getBCRs(node, false)}) {
```

Comment:
this is a nice little generic helper! might be nice in `page-functions.js`

Reply:
we've also been trying to keep all our in-page functions there, but we're going to have so many for this function it's probably worth creating a dedicated file for them. 

do you think it's easiest for it to be basically one giant in-page evaluation?

I ask because it's generally hard to debug and figure out what's going on in the in-page functions than the node side, but it looks like you've worked out a decent method of development here with the `fs` call. There might also be some opportunity to share some BCR logic between audit and gatherer?

Reply:
We've mostly been using a [manually synced gist](https://gist.github.com/mattzeunert/67dcea9c97a882248508a81e11adaf4a) for development. It also gives a visual output, which we're hoping to add to the DevTools Audit panel one day.

Will move the BCR logic to a shared file.